test/unit/puppetlabs/jruby_utils/lockable_pool_test.clj

```diff
@@ -591,17 +591,81 @@
       ; Pool is empty and the borrow future is blocked
-      (is (not (realized? borrow-promise)))
+      (is (not (realized? blocked-borrow)))
       ; inserting the pill should unblock the promise
       (.insertPill pool pill)
       ; borrow finishes and gives us back the pill
-      (let [promise-result (timed-deref borrow-promise)]
-        (is (identical? promise-result pill))))))
+      (let [future-result (timed-deref blocked-borrow)]
+        (is (identical? future-result pill)))))
+
+  (testing "second insert doesn't change the pill"
+    (let [pool (create-populated-pool 1)
+          first-pill (str "pill clinton")
+          second-pill (str "pillary clinton")]
+      (.insertPill pool first-pill)
+      (is (identical? first-pill (.borrowItem pool)))
+
+      (.insertPill pool second-pill)
+      ; Should still be equal to the first pill
+      (is (identical? first-pill (.borrowItem pool)))
+      (is (not (identical? second-pill (.borrowItem pool)))))))
+
+(deftest release-item-exceptions-test
+  (testing "releasing a different pill than the one that was inserted errors"
+    (let [pool (create-populated-pool 1)
+          first-pill (str "a city upon a pill")
+          second-pill (str "capitol pill")]
+      (.insertPill pool first-pill)
+      ; Only to show that it does not error
+      (is (nil? (.releaseItem pool first-pill)))
+      (is (thrown-with-msg?
+           IllegalArgumentException
+           #"The item being released is not registered with the pool"
+           (.releaseItem pool second-pill)))))
+
+  (testing "releasing a jruby not registered with the pool errors"
+    (let [pool (create-populated-pool 1)
+          not-in-pool-instance "I was never registered"]
+      (is (thrown-with-msg?
+           IllegalArgumentException
+           #"The item being released is not registered with the pool"
+           (.releaseItem pool not-in-pool-instance))))))
+
+(deftest lock-interrupted-by-pill-insertion-test
+  (testing "a call to .lock will throw if there is a pill"
+    (let [pool (create-populated-pool 1)
+          pill "pilliam shakespeare"]
+      (.insertPill pool pill)
+      (is (thrown-with-msg?
+           InterruptedException
+           #"Lock can't be granted because a pill has been inserted"
+           (.lock pool)))))
+
+  (testing "a blocked .lock call throws an InterruptedException once a pill is inserted"
+    (let [pool (create-populated-pool 1)
+          pill "pillful ignorance"
+          instance (.borrowItem pool)]
+      ; Exceptions thrown from the future will be returned as InterruptedException,
+      ; so we can't use thrown-with-msg?. We'll catch it, return it instead of
+      ; throwing it, and inspect it manually below
+      (let [blocked-lock-future (future (try (.lock pool)
+                                             (catch InterruptedException e
+                                               e)))]
+        ; Give future time to start
+        (Thread/sleep 500)
```

Comment:
To avoid the sleep here (and make this a bit less racy), I think you could wait on an `.isLocked()` call for the pool to return true before continuing.  That should happen once the pool lock has been reserved but before the wait on the `lockAvailable` condition, so seems like it would still cover what we want to test here.